pyccel/ast/builtins.py

```diff
@@ -468,11 +478,14 @@ def __init__(self, *args):
                 self._rank  = len(self._shape)
 
         else:
-            self._rank      = max(a.rank for a in args) + 1
+            max_rank = max(a.rank for a in args)
+            self._rank      = max_rank + 1
             self._dtype     = NativeGeneric()
             self._precision = 0
             if self._rank == 1:
                 self._shape     = (LiteralInteger(len(args)), )
+            elif any(a.rank != max_rank for a in args):
+                self._shape     = (LiteralInteger(len(args)), ) + (None,)*(self._rank-1)
             else:
                 self._shape     = (LiteralInteger(len(args)), ) + args[0].shape
```

Comment:
Would this not work then?
```suggestion
            else:
                extra_dims  = args[0].shape or (None,) * (self._rank - 1)
                self._shape = (LiteralInteger(len(args)), ) + extra_dims
```

Reply:
I don't think so. If we have:
```python
a = ((1,2), ((3,4),(5,6))
```
Then your solution will give:
```
rank = 3
shape = (2,2)
```
whereas mine will give:
```
rank = 3
shape = (2,None,None)
```

Maybe the mismatch between the rank and the shape would be enough to trigger a neat error but I am not sure. This would need to be tested

Reply:
I've just checked. Your solution raises an `AssertionError` with the example I wrote, while the implemented error raises a neat `NotImplementedError`:
```
ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: TUPLE_WITH_INCOHERENT_RANKS.py [3,4]| Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language
Cannot infer shape of right-hand side for expression a = ((1, 2), ((1, 2), (3, 4)))
```

Reply:
There is a missing parenthesis in your example, or an extra one.

Reply:
Oops. Fixed

Reply:
All right, I see the problem now!